### PR TITLE
cli: refactors; make `cockroach-sql`'s CLI parameter handling match `cockroach sql`

### LINKED ISSUE #81882: Allow standalone SQL client binary to support --certs-dir
The standalone SQL client binary does not currently support --certs-dir, meaning you have to use --url to pass an unwield postgresql-type URL rather than being able to simply run the binary (and it finding .cockroach-certs/ by default or whatever is passed by --certs-dir).

Jira issue: CRDB-16106

### LINKED ISSUE #82024: cli: standalone `cockroach-sql` does not know how to use env vars
**Describe the problem**

`cockroach sql` supports COCKROACH_DATABASE and other client-side env vars to set connection defaults.

The standalone `cockroach-sql` shell does not see these env vars and so the env vars cannot be used to override the connection parameters.

**Expected behavior**

The env vars should be supported for convenience and feature parity with `cockroach sql`.

Jira issue: CRDB-16172

### PR DESCRIPTION
**Pull request by knz**
Note to the reviewers:
- The PR has been split into many small commits to ease the review.
- The "main commit" is the last in the sequence; all commits before the last constitute a long-awaited refactor of the connection parameter handling, including moving much of the security-related logic away from `cli` into a sub-package of `security` where it belongs.
- We only have testing code to exercise the command-line handling for `cockroach sql`, which at this point should be considered sufficient given we're using the same config code in `cockroach-sql` now. To make tests work across the two commands, we would need to make progress on #80921 first.

Fixes #81882.
Fixes #82024.
Unblocks progress on #29285.

### CODE REVIEW

**Comment by knz** [comment]
NB: Reviewable seems to have issues displaying the entire commit sequence. The Github review interface is fine with it.

If the reviewer(s) would prefer that, I can split this PR into a sequence of small(er) PRs.

File: pkg/cli/client_url.go [hunk 1]
```diff
@@ -160,6 +155,74 @@ func (u urlParser) setInternal(v string, warn bool) error {
...
+// sslStrict, when set to true, requires that the SSL file paths in
+// a URL clearly map to a single certificate directory and restricts
+// the set of supported SSL modes to just "disable" and "require".
+//
+// This is set for all non-SQL client commands, which only support
+// the insecure boolean and certs-dir with maximum SSL validation.
+func updateURL(
+	newURL string,
+	sslStrict bool,
+	makeStrictErr func() error,
```
Comment (otan) [inline pkg/cli/client_url.go]: optional: how do you feel about these functions being an interface instead? 
could add unit tests too, if we wanted to. having to pass a lot of hooks in is easy to lose track of.

File: pkg/cmd/cockroach-sql/main.go [hunk 2]
```diff
@@ -102,14 +152,45 @@ var sqlCmd = &cobra.Command{
...
-	if url == "" {
-		return errors.New("no connection URL specified")
+
+	// Translate the connection arguments to a connection URL.
+	connURL, err := clientconnurl.MakeClientConnURL(copts)
+	if err != nil {
+		return err
+	}
+
+	// How we're going to authenticate.
```
Comment (otan) [inline pkg/cmd/cockroach-sql/main.go:187]: nit: comment sounds like a question - maybe "Get how we're going to authenticate"

**Review by otan — COMMENTED** [review]
mostly some naming nits here, curious about how you feel about them

the `clientconnurl` package and `clienturl` package is a little similarly named and easy to get confused by. what are your thoughts on:

* clientconnurl -> pgconnurl (is it client specific?)
* clienturl -> cliurl, to make it clear only the CLI should be using it?

**Comment by knz** [comment]
Yes I agree some names here are unfortunate.

I will split this PR so we can iterate on the review more easily.

**Comment by knz** [comment]
I've split the PR into the following pieces:
- #82068
- #82069 
- #82070 
- #82071 
- #82072 
- #82074

File: pkg/security/securityassets/security_assets.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+package securityassets
+
+import (
+	"io/ioutil"
+	"os"
+)
+
+// AssetLoader describes the functions necessary to read certificate and key files.
+type AssetLoader struct {
```
[the remark this answers is not in the document]
Comment (knz) [reply pkg/security/securityassets/security_assets.go:21]: done

File: pkg/security/certnames/certs_locator.go [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+import (
+	"os"
+	"path/filepath"
+
+	"github.com/cockroachdb/cockroach/pkg/security/username"
+	"github.com/cockroachdb/errors/oserror"
+)
+
+// A CertsLocator provides locations to certificates.
+type CertsLocator struct {
```
[the remark this answers is not in the document]
Comment (knz) [reply pkg/security/certnames/certs_locator.go:21]: agreed. Done (in split PR)

File: pkg/cli/client_url.go (hunk 1 above)
Comment (knz) [reply pkg/cli/client_url.go]: good idea. done (in split PR)

**Comment by knz** [comment]
unrelated flake

bors r=otan,catj-cockroach

**Comment by knz** [comment]
added a skip

bors r=otan,catj-cockroach

**Pull request merged**